# Assignment 3 - Is a Picture Worth a Thousand Words?

Team Members:
- Abigail Peck (ap69393)
- Emily Caraher (ec45335)
- Nandini Anand Kumar (na29337)
- Riju Hariharan (rh42988)
- Sarvesh Miskin (sm88728)
- Varsha Ramesh (vr23656)

You are an analytics consultant to a non-profit organization (e.g., a pet rescue society, a children’s hospital, etc.). Your objective is to help the organization raise more funds in its gofundme.com campaigns.\
\
Task A: Scrape ~1000 fundraising campaigns in a particular category (e.g., Golden Retriever rescue society, an organization raising funds for wildfire victims, etc.). Fetch (i) image URLs, (ii) text description of a post, (iii) $ raised, (iv) how long the campaign has been running.

In [ ]:
# """
# GoFundMe Business Category Scraper
# Scrapes campaigns from https://www.gofundme.com/discover/business-fundraiser
# """

# from selenium import webdriver
# from selenium.webdriver.common.by import By
# from selenium.webdriver.chrome.options import Options
# import pandas as pd
# import time
# import re
# from datetime import datetime
# from bs4 import BeautifulSoup
# from selenium.webdriver.support.ui import WebDriverWait
# from selenium.webdriver.support import expected_conditions as EC

# class GoFundMeBusinessScraper:
#     def __init__(self, headless=True):
#         """Initialize the scraper with Selenium"""
#         chrome_options = Options()
#         if headless:
#             chrome_options.add_argument("--headless")
#         chrome_options.add_argument("--no-sandbox")
#         chrome_options.add_argument("--disable-dev-shm-usage")
#         chrome_options.add_argument("--disable-blink-features=AutomationControlled")
#         chrome_options.add_argument("--disable-gpu")
#         chrome_options.add_argument("--window-size=1920,1080")
#         chrome_options.add_argument("--disable-extensions")
#         chrome_options.add_argument("--proxy-server='direct://'")
#         chrome_options.add_argument("--proxy-bypass-list=*")
#         chrome_options.add_argument("--start-maximized")
#         chrome_options.add_argument("user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")

#         # Prevent crashes
#         chrome_options.add_argument("--disable-software-rasterizer")
#         chrome_options.add_argument("--disable-dev-tools")
#         chrome_options.add_experimental_option('excludeSwitches', ['enable-logging'])

#         self.driver = webdriver.Chrome(options=chrome_options)
#         self.driver.set_page_load_timeout(30)
#         self.campaigns = []
#         self.base_url = "https://www.gofundme.com/discover/business-fundraiser"

#     def click_show_more(self, max_clicks=50):
#         """Click the 'Show More' button to load more campaigns"""
#         clicks = 0
#         while clicks < max_clicks:
#             try:
#                 # Scroll to bottom first to make button visible
#                 self.driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
#                 time.sleep(1)

#                 # Find and click the show more button
#                 show_more_button = self.driver.find_element(
#                     By.CSS_SELECTOR,
#                     "button.hrt-secondary-button.hrt-base-button"
#                 )

#                 # Scroll button into view and click
#                 self.driver.execute_script("arguments[0].scrollIntoView(true);", show_more_button)
#                 time.sleep(1)
#                 show_more_button.click()

#                 clicks += 1
#                 print(f"Clicked 'Show More' button {clicks}/{max_clicks} times")
#                 time.sleep(2)  # Wait for content to load

#             except Exception as e:
#                 print(f"No more 'Show More' button found or error: {str(e)}")
#                 break

#         print(f"Finished loading campaigns after {clicks} clicks")

#     def scrape_campaign_cards(self, max_campaigns=1000):
#         """Scrape campaign cards from the discover page"""
#         print(f"Loading {self.base_url}")
#         self.driver.get(self.base_url)

#         # Wait for page to load
#         time.sleep(5)

#         # Click "Show More" button to load more campaigns
#         # Starts with 24 campaigns, adds 12 per click
#         # Formula: 24 + (12 * num_clicks) >= max_campaigns
#         # num_clicks = (max_campaigns - 24) / 12
#         num_clicks = max(0, int((max_campaigns - 24) / 12) + 2)  # +2 for buffer
#         print(f"Need approximately {num_clicks} clicks to load {max_campaigns} campaigns")
#         self.click_show_more(max_clicks=num_clicks)

#         # Find all campaign cards using the specific class
#         print("Finding campaign cards...")
#         campaign_elements = self.driver.find_elements(By.CSS_SELECTOR, "a.hrt-action-card.hrt-base-button")
#         print(f"Found {len(campaign_elements)} campaigns")

#         # Extract unique campaign URLs
#         campaign_urls = []
#         for elem in campaign_elements:
#             try:
#                 url = elem.get_attribute('href')
#                 if url and '/f/' in url and url not in campaign_urls:
#                     campaign_urls.append(url)
#             except:
#                 continue

#         print(f"Processing {min(len(campaign_urls), max_campaigns)} unique campaigns")

#         # Scrape each campaign (limit to max_campaigns)
#         for i, url in enumerate(campaign_urls[:max_campaigns]):
#             print(f"Scraping campaign {i+1}/{min(len(campaign_urls), max_campaigns)}: {url}")

#             # Try scraping with retry logic
#             retries = 3
#             for attempt in range(retries):
#                 try:
#                     campaign_data = self.scrape_campaign_page(url)
#                     if campaign_data:
#                         self.campaigns.append(campaign_data)
#                     break
#                 except Exception as e:
#                     print(f"Attempt {attempt+1} failed: {str(e)}")
#                     if attempt < retries - 1:
#                         print("Retrying...")
#                         time.sleep(3)
#                     else:
#                         print(f"Skipping campaign after {retries} attempts")

#             time.sleep(2)  # Be polite

#         return self.campaigns

#     def scrape_campaign_page(self, url):
#         """Scrape individual campaign page with reliable image and full description support"""
#         try:
#             # Restart driver if crashed
#             try:
#                 _ = self.driver.current_url
#             except:
#                 print("Browser crashed, reinitializing...")
#                 self.__init__(headless=True)
#                 time.sleep(2)

#             self.driver.get(url)
#             time.sleep(3)

#             campaign_data = {
#                 'url': url,
#                 'title': None,
#                 'image_url': None,
#                 'description': None,
#                 'amount_raised': None,
#                 'days_running': None
#             }

#             # Title
#             try:
#                 h1 = next(h for h in self.driver.find_elements(By.TAG_NAME, "h1") if h.text.strip())
#                 campaign_data['title'] = h1.text.strip()
#             except:
#                 pass

#             # Image
#             try:
#                 img = self.driver.find_element(By.TAG_NAME, 'img')
#                 campaign_data['image_url'] = img.get_attribute('src')
#             except:
#                 print(f"Image scrape failed")

#             # Full description from nested divs inside campaign-description container
#             try:
#                 desc_root = self.driver.find_element(By.CSS_SELECTOR, 'div.campaign-description_campaignDescription__6P_RU')
#                 html = desc_root.get_attribute('innerHTML')
#                 soup = BeautifulSoup(html, 'html.parser')

#                 # Remove the "Read more" button to exclude its text
#                 button = soup.find('button', {'data-element-id': 'btn_story_read-more'})
#                 if button:
#                     button.decompose()

#                 # Replace <br> tags with newlines
#                 for br in soup.find_all("br"):
#                     br.replace_with("\n")

#                 # Extract only visible text, ignoring nested duplication
#                 description = soup.get_text(separator="\n", strip=True)

#                 campaign_data['description'] = description
#             except Exception as e:
#                 print(f"Description scrape failed: {e}")

#             # Amount raised
#             try:
#                 body_text = self.driver.find_element(By.TAG_NAME, 'body').text
#                 match = re.search(r'\$([0-9,]+)\s*raised', body_text, re.IGNORECASE)
#                 if match:
#                     campaign_data['amount_raised'] = match.group(1).replace(',', '')
#             except:
#                 pass

#             # Created date and days running
#             try:
#                 date_elem = self.driver.find_element(By.CSS_SELECTOR, 'span.a-created-date')
#                 raw_date = date_elem.text.strip().replace('Created', '').strip()
#                 campaign_data['days_running'] = self.calculate_days_since(raw_date)
#             except:
#                 pass

#             return campaign_data

#         except Exception as e:
#             print(f"Error scraping {url}: {e}")
#             return None

#     def calculate_days_since(self, date_text):
#         """Calculate days since a given date string or relative phrase"""
#         try:
#             text = date_text.lower()

#             # Handle relative formats
#             if 'hour' in text and 'ago' in text:
#                 match = re.search(r'(\d+)\s*hours?\s*ago', text, re.IGNORECASE)
#                 if match:
#                     return int(match.group(1)) / 24  # Convert hours to days

#             if ('day' in text or 'd ago' in text) and 'ago' in text:
#                 match = re.search(r'(\d+)\s*d(?:ays?)?\s*ago', text, re.IGNORECASE)
#                 if match:
#                     return int(match.group(1))

#             if 'month' in text and 'ago' in text:
#                 match = re.search(r'(\d+)\s*months?\s*ago', text, re.IGNORECASE)
#                 if match:
#                     return int(match.group(1)) * 30

#             if 'year' in text and 'ago' in text:
#                 match = re.search(r'(\d+)\s*years?\s*ago', text, re.IGNORECASE)
#                 if match:
#                     return int(match.group(1)) * 365

#             # Handle absolute date like "September 11th, 2025"
#             # Remove ordinal suffixes: st, nd, rd, th
#             clean_date = re.sub(r'(\d+)(st|nd|rd|th)', r'\1', date_text)

#             # Try parsing full date format
#             try:
#                 date_obj = datetime.strptime(clean_date.strip(), '%B %d, %Y')
#             except ValueError:
#                 # Try alternative format if needed
#                 date_obj = datetime.strptime(clean_date.strip(), '%b %d, %Y')  # e.g., "Sep 11, 2025"

#             today = datetime.now()
#             return (today - date_obj).days

#         except Exception as e:
#             print(f"Could not parse date: {date_text}, error: {e}")
#             return None

#     def save_to_csv(self, filename='business_campaigns.csv'):
#         """Save to CSV"""
#         df = pd.DataFrame(self.campaigns)
#         df.to_csv(filename, index=False)
#         print(f"\nSaved {len(self.campaigns)} campaigns to {filename}")
#         return df

#     def close(self):
#         """Close the browser"""
#         self.driver.quit()

# # Usage
# if __name__ == "__main__":
#     # Create scraper instance
#     scraper = GoFundMeBusinessScraper(headless=False)  # Set to True to hide browser
#     try:
#         # Scrape detailed campaign data by visiting each page
#         print("Scraping detailed campaign data by visiting each page...")
#         campaigns = scraper.scrape_campaign_cards(max_campaigns=1000)

#         # Save to CSV
#         df = scraper.save_to_csv('business_campaigns.csv')

#         # Display summary
#         print(f"\n{'='*50}")
#         print(f"Total campaigns scraped: {len(campaigns)}")
#         print(f"{'='*50}")
#     finally:
#         scraper.close()

Task B: Using the image URLs, obtain image labels (text) from Google Vision (cloud service) or other services such as Azure. You can also use an LLM through its API. You will need an account, though.

In [ ]:
# pip install pandas requests

In [ ]:
# import pandas as pd
# import requests
# import json
# import time

# API_KEY = "AIzaSyD2MAHZ5gmPMN_SL06zxHBylAIJdfKJ_Z0"  # Replace with your actual API key
# VISION_URL = f"https://vision.googleapis.com/v1/images:annotate?key=AIzaSyD2MAHZ5gmPMN_SL06zxHBylAIJdfKJ_Z0"

# # Load your CSV
# df = pd.read_csv("business_campaigns.csv")

# labels_list = []

# for idx, url in enumerate(df['image_url']):
#     print(f"Processing image {idx+1}/{len(df)}")
#     # Build request payload
#     request_body = {
#         "requests": [
#             {
#                 "image": {
#                     "source": {
#                         "imageUri": url
#                     }
#                 },
#                 "features": [
#                     {
#                         "type": "LABEL_DETECTION",
#                         "maxResults": 5
#                     }
#                 ]
#             }
#         ]
#     }

#     try:
#         response = requests.post(VISION_URL, json=request_body)
#         result = response.json()

#         # Extract labels
#         labels = result['responses'][0].get('labelAnnotations', [])
#         label_descriptions = [label['description'] for label in labels]
#         labels_list.append(", ".join(label_descriptions))

#         # Respect API limits
#         time.sleep(0.2)

#     except Exception as e:
#         print(f"Error processing image: {e}")
#         labels_list.append("")

# # Add results to dataframe
# df['labels'] = labels_list

# # Save to new CSV
# df.to_csv("labeled_images.csv", index=False)

In [ ]:
# from google.colab import files
# files.download("labeled_images.csv")

Upload labeled_images.csv here - contains the following columns: url, title, image_url, description, amount_raised, days_running, and labels.

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving labeled_images.csv to labeled_images.csv


Task C: Create a column called binary (lowercase only) where value =1 (stands for high $$) or 0 (stands for low $) based on the median value.

In [1]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.preprocessing import StandardScaler
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score
from scipy.sparse import hstack
from sklearn.feature_extraction.text import TfidfVectorizer

C:\Users\kevin\anaconda3\lib\site-packages\pandas\core\arrays\masked.py:61: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


In [2]:
df_labeled = pd.read_csv('labeled_images.csv')
df_labeled['amount_raised'] = pd.to_numeric(df_labeled['amount_raised'], errors='coerce')
median_raised = df_labeled['amount_raised'].median()
df_labeled['binary'] = df_labeled['amount_raised'].apply(lambda x: 1 if x >= median_raised else 0)

Task D: Run a logistic regression with binary as the dependent variable, and the image_labels as independent variables. You can use a BoW model for text. Figure out how you can use the duration of the campaign in the model. What is the accuracy (show the confusion matrix) of this prediction model? The idea is to be able to predict the engagement level for an image.\
\
Accuracy = 1 - # prediction errors / total # cases\
\
What accuracy do you get by using the description words as the independent variables instead of image_labels? Finally, what accuracy do you get by combining (concatenating) the image_labels and word descriptions and using them together as independent variables? What can you conclude from your analysis?

In [3]:
#bag of words for image labels
df_labeled['labels'] = df_labeled['labels'].fillna("") #if no image labels changes from nan to an empty string
vectorizer_img = CountVectorizer()
X_img = vectorizer_img.fit_transform(df_labeled['labels'])

#bag of words for description text
vectorizer_txt = CountVectorizer(max_features=1000, stop_words='english')
X_text = vectorizer_txt.fit_transform(df_labeled['description'].fillna('')) # Fill NaN descriptions

#Scale and Normalize days running of each fundraiser
# Fill NaN values in 'days_running' with the median before scaling
df_labeled['days_running'] = df_labeled['days_running'].fillna(df_labeled['days_running'].median())
X_duration = np.array(df_labeled['days_running']).reshape(-1, 1)
scaler = StandardScaler()
X_duration_scaled = scaler.fit_transform(X_duration)

#dependent variable
y = df_labeled['binary']

#split the train and testing data
#only split the data once and use the same indices for each independent variable
X_train_idx, X_test_idx = train_test_split(df_labeled.index, test_size=0.2, random_state=42)
X_train_img, X_test_img = X_img[X_train_idx], X_img[X_test_idx]
X_train_txt, X_test_txt = X_text[X_train_idx], X_text[X_test_idx]
X_train_dur, X_test_dur = X_duration_scaled[X_train_idx], X_duration_scaled[X_test_idx]
y_train, y_test = y[X_train_idx], y[X_test_idx]


# Combine sparse and dense features
X_train_img_full = hstack([X_train_img, X_train_dur])
X_test_img_full  = hstack([X_test_img, X_test_dur])

logreg = LogisticRegression(max_iter=5000)
logreg.fit(X_train_img_full, y_train)

y_pred_img = logreg.predict(X_test_img_full)
acc_img = accuracy_score(y_test, y_pred_img)
cm_img = confusion_matrix(y_test, y_pred_img)

print("Image-label + Duration model accuracy:", round(acc_img, 3))
print("Confusion matrix:\n", cm_img)


# Model with Description Text + Days Running
X_train_txt_full = hstack([X_train_txt, X_train_dur])
X_test_txt_full  = hstack([X_test_txt, X_test_dur])

logreg_txt = LogisticRegression(max_iter=5000)
logreg_txt.fit(X_train_txt_full, y_train)

y_pred_txt = logreg_txt.predict(X_test_txt_full)
acc_txt = accuracy_score(y_test, y_pred_txt)
cm_txt = confusion_matrix(y_test, y_pred_txt)

print("\nDescription-text + Duration model accuracy :", round(acc_txt, 3))
print("Confusion matrix:\n", cm_txt)

# Model with Combined Features (Image Labels + Description Text + Days Running)
X_train_combined = hstack([X_train_img, X_train_txt, X_train_dur])
X_test_combined  = hstack([X_test_img, X_test_txt, X_test_dur])

logreg_combined = LogisticRegression(max_iter=5000)
logreg_combined.fit(X_train_combined, y_train)

y_pred_combined = logreg_combined.predict(X_test_combined)
acc_combined = accuracy_score(y_test, y_pred_combined)
cm_combined = confusion_matrix(y_test, y_pred_combined)

print("\n Image-label + Description-text + Duration model accuracy:", round(acc_combined, 3))
print("Confusion matrix:\n", cm_combined)

print("\n Without Duration")
logreg_img_only = LogisticRegression(max_iter=5000)
logreg_img_only.fit(X_train_img, y_train)

# --- Predict and Evaluate ---
y_pred_img_only = logreg_img_only.predict(X_test_img)
acc_img_only = accuracy_score(y_test, y_pred_img_only)
cm_img_only = confusion_matrix(y_test, y_pred_img_only)

# --- Print Results ---
print("\n--- Model: Image Labels ONLY ---")
print("Accuracy:", round(acc_img_only, 4))
print("Confusion matrix:\n", cm_img_only)

logreg_txt_only = LogisticRegression(max_iter=5000)
logreg_txt_only.fit(X_train_txt, y_train)

# --- Predict and Evaluate ---
y_pred_txt_only = logreg_txt_only.predict(X_test_txt)
acc_txt_only = accuracy_score(y_test, y_pred_txt_only)
cm_txt_only = confusion_matrix(y_test, y_pred_txt_only)

# --- Print Results ---
print("\n--- Model: Description Text ONLY ---")
print("Accuracy:", round(acc_txt_only, 4))
print("Confusion matrix:\n", cm_txt_only)

X_train_bow_only = hstack([X_train_img, X_train_txt])
X_test_bow_only  = hstack([X_test_img, X_test_txt])

# --- Run Logistic Regression ---
# Using max_iter=5000 for stability with large sparse feature matrices
logreg_bow_only = LogisticRegression(max_iter=5000)
logreg_bow_only.fit(X_train_bow_only, y_train)

# --- Predict and Evaluate ---
y_pred_bow_only = logreg_bow_only.predict(X_test_bow_only)
acc_bow_only = accuracy_score(y_test, y_pred_bow_only)
cm_bow_only = confusion_matrix(y_test, y_pred_bow_only)

# --- Print Results ---
print("\n--- Model: Image Labels + Description Text ONLY ---")
print("Accuracy:", round(acc_bow_only, 4))
print("Confusion matrix:\n", cm_bow_only)

Image-label + Duration model accuracy: 0.735
Confusion matrix:
 [[82 23]
 [30 65]]

Description-text + Duration model accuracy : 0.735
Confusion matrix:
 [[81 24]
 [29 66]]

 Image-label + Description-text + Duration model accuracy: 0.74
Confusion matrix:
 [[82 23]
 [29 66]]

 Without Duration

--- Model: Image Labels ONLY ---
Accuracy: 0.72
Confusion matrix:
 [[88 17]
 [39 56]]

--- Model: Description Text ONLY ---
Accuracy: 0.695
Confusion matrix:
 [[80 25]
 [36 59]]

--- Model: Image Labels + Description Text ONLY ---
Accuracy: 0.69
Confusion matrix:
 [[78 27]
 [35 60]]


The accuracy for each of the first three logistic regressions we did was almost identical. We added duration in as a factor for each of the first three models for uniformity. But, due to the almost identical accuracies we believed that the duration of the business fundraiser held too much weight. To verify our assumption we performed logistic regression again on each model but this time without factoring in the duration of the fundraiser.

All three new models (without duration) performed slightly worse, with the description text model only obtaining an accuracy of 69.5%. And, the image labels + description text model getting the lowest accuracy of 69%. This proved that duration was the factor in our model that was leading us to get such similar accuracy scores. In the image label only model the accuracy was still quite close to the duration models, but when paired with description text the accuracy dropped. We believe this means description text is the worst at predicting how much a fundraiser will make, and image labels do a slightly better job.

Overall, the duration of a fundraiser does the best at predicting how much each fundraiser will make. We got our best prediction scores when we used duration with image labels, and all three factors together. In order to make the most money for your fundraiser, you most likely need to have it running for a longer period of time. Images also are more likely to inspire donations from your audience than your description of your fundraiser.

Task E: Perform topic modeling (LDA) on the original image_labels or word descriptions depending on which category gave higher accuracy in D. If the combination of image_labels and description words gave the highest accuracy, use them together. Choose an appropriate number of topics. You may want to start with 4-5 topics, but adjust the number up or down depending on the word distributions you get. Decide on suitable names for each topic. Now sort the data from high to low \$ raised (don’t use the binary column, use the actual \$), and consider the highest and the lowest quartiles of $. What are the main differences in the average topic weights of images across the two quartiles (e.g., greater weight of some topics in the highest versus lowest quartiles)? Show the main results in a table.

In [4]:
# --- Task E: Topic Modeling + Quartile Comparison ---

import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation

# 1️⃣ Combine text fields (if both labels and descriptions exist)
df_labeled['labels'] = df_labeled['labels'].fillna("")
df_labeled['description'] = df_labeled['description'].fillna("")
df_labeled['combined_text'] = df_labeled['labels'] + " " + df_labeled['description']

# 2️⃣ Bag-of-Words representation
vectorizer = CountVectorizer(stop_words='english')
X = vectorizer.fit_transform(df_labeled['combined_text'])
words = np.array(vectorizer.get_feature_names_out())

# 3️⃣ Fit LDA model (5 topics)
lda = LatentDirichletAllocation(n_components=5, random_state=42)
topic_matrix = lda.fit_transform(X)

# 4️⃣ Save word–topic probabilities (for naming topics)
topic_word_matrix = lda.components_
topic_word_probs = topic_word_matrix / topic_word_matrix.sum(axis=1)[:, np.newaxis]
word_topic_matrix = pd.DataFrame(topic_word_probs.T,
                                 columns=[f"Topic_{i+1}" for i in range(lda.n_components)])
word_topic_matrix.insert(0, "word", words)
word_topic_matrix.to_csv("TaskE_word_topic_matrix3.csv", index=False)
print("✅ Saved 'TaskE_word_topic_matrix.csv' (word–topic probabilities)")

# 5️⃣ Quartile comparison on amount_raised
topic_cols = [f"Topic_{i+1}" for i in range(lda.n_components)]
df_topics = pd.concat([df_labeled[['amount_raised']].reset_index(drop=True),
                       pd.DataFrame(topic_matrix, columns=topic_cols)], axis=1)

# Normalize topic weights per campaign
df_topics[topic_cols] = df_topics[topic_cols].div(df_topics[topic_cols].sum(axis=1), axis=0)

# Split into quartiles by actual $ raised
df_sorted = df_topics.sort_values('amount_raised', ascending=False)
q1 = df_sorted['amount_raised'].quantile(0.25)
q3 = df_sorted['amount_raised'].quantile(0.75)
top_quartile = df_sorted[df_sorted['amount_raised'] >= q3]
bottom_quartile = df_sorted[df_sorted['amount_raised'] <= q1]

# Calculate average topic weights and differences
top_avg = top_quartile[topic_cols].mean()
bottom_avg = bottom_quartile[topic_cols].mean()
topic_comparison = pd.DataFrame({
    'High $ (Top 25%)': top_avg,
    'Low $ (Bottom 25%)': bottom_avg
})
topic_comparison['Difference'] = topic_comparison['High $ (Top 25%)'] - topic_comparison['Low $ (Bottom 25%)']
topic_comparison = topic_comparison.sort_values('Difference', ascending=False)

# Save only the comparison table
topic_comparison.to_csv("TaskF_topic_quartile_comparison3.csv")
print("\n✅ Saved 'TaskF_topic_quartile_comparison.csv'")
print(topic_comparison.round(3))


✅ Saved 'TaskE_word_topic_matrix.csv' (word–topic probabilities)

✅ Saved 'TaskF_topic_quartile_comparison.csv'
         High $ (Top 25%)  Low $ (Bottom 25%)  Difference
Topic_3             0.368               0.064       0.304
Topic_2             0.369               0.119       0.250
Topic_5             0.179               0.076       0.103
Topic_1             0.030               0.035      -0.005
Topic_4             0.053               0.706      -0.653


Task F: What advice would you give to the organization if it wants to increase the $ raised based on your findings?

In [5]:
# --- Task E: Topic Modeling (LDA) on Image Labels ---

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation
import pandas as pd
import numpy as np

# 1️⃣ Vectorize the image labels
# (We reuse df_labeled['labels'] from previous tasks)
vectorizer_lda = CountVectorizer(stop_words='english')
X_lda = vectorizer_lda.fit_transform(df_labeled['labels'].fillna(""))

# 2️⃣ Run LDA (start with 5 topics — adjust if needed)
lda = LatentDirichletAllocation(n_components=5, random_state=42)
topic_matrix = lda.fit_transform(X_lda)   # Each row = topic distribution for one campaign

# 3️⃣ Add topic weights to the dataframe
topic_cols = [f"Topic_{i+1}" for i in range(lda.n_components)]
df_topics = pd.concat([df_labeled.reset_index(drop=True),
                       pd.DataFrame(topic_matrix, columns=topic_cols)], axis=1)

# 4️⃣ Sort by amount raised and split into top/bottom quartiles
df_sorted = df_topics.sort_values('amount_raised', ascending=False)
q1 = df_sorted['amount_raised'].quantile(0.25)
q3 = df_sorted['amount_raised'].quantile(0.75)

top_quartile = df_sorted[df_sorted['amount_raised'] >= q3]
bottom_quartile = df_sorted[df_sorted['amount_raised'] <= q1]

# 5️⃣ Compute average topic weights for each quartile
top_avg = top_quartile[topic_cols].mean()
bottom_avg = bottom_quartile[topic_cols].mean()

# 6️⃣ Create a comparison table
topic_comparison = pd.DataFrame({
    'High $ (Top 25%)': top_avg,
    'Low $ (Bottom 25%)': bottom_avg
})
topic_comparison['Difference'] = topic_comparison['High $ (Top 25%)'] - topic_comparison['Low $ (Bottom 25%)']

print("\n📊 Average Topic Weights by Quartile:")
print(topic_comparison.round(3))

# 7️⃣ Display top keywords for each topic
words = np.array(vectorizer_lda.get_feature_names_out())
for i, topic in enumerate(lda.components_):
    top_words = [words[j] for j in topic.argsort()[-10:]]
    print(f"\n🔹 Topic {i+1} top words: {top_words}")



📊 Average Topic Weights by Quartile:
         High $ (Top 25%)  Low $ (Bottom 25%)  Difference
Topic_1             0.156               0.196      -0.039
Topic_2             0.262               0.217       0.045
Topic_3             0.300               0.213       0.087
Topic_4             0.152               0.195      -0.043
Topic_5             0.130               0.179      -0.049

🔹 Topic 1 top words: ['retail', 'text', 'finger', 'design', 'shelf', 'customer', 'furniture', 'restaurant', 'beauty', 'food']

🔹 Topic 2 top words: ['red', 'wood', 'art', 'symbol', 'advertising', 'font', 'logo', 'graphic', 'graphics', 'design']

🔹 Topic 3 top words: ['chin', 'glasses', 'eyewear', 'care', 'hair', 'eyebrow', 'expression', 'facial', 'happiness', 'smile']

🔹 Topic 4 top words: ['arts', 'animation', 'car', 'cartoon', 'carnivores', 'motor', 'tire', 'automotive', 'blue', 'vehicle']

🔹 Topic 5 top words: ['brake', 'aircraft', 'electronic', 'camera', 'hat', 'cap', 'night', 'food', 'light', 'automot

In [6]:
words

array(['absorber', 'academic', 'active', 'advertising', 'aerospace',
       'afro', 'agriculture', 'air', 'aircraft', 'album', 'alcoholic',
       'aluminium', 'american', 'animal', 'animated', 'animation',
       'appliance', 'apron', 'area', 'arm', 'art', 'artist', 'arts',
       'asphalt', 'audio', 'automotive', 'aviation', 'awning', 'badge',
       'bag', 'bakery', 'baking', 'balance', 'ball', 'ballet', 'balloon',
       'bank', 'banner', 'bar', 'barefoot', 'barrel', 'barware',
       'baseball', 'basket', 'beard', 'beauty', 'bed', 'bedrock', 'bench',
       'berry', 'bicycle', 'billboard', 'bird', 'black', 'blond',
       'blossom', 'blue', 'boat', 'body', 'bone', 'book', 'bookcase',
       'bottle', 'bowl', 'brake', 'brambles', 'branch', 'bread',
       'brewery', 'brick', 'brickwork', 'bridle', 'brown', 'building',
       'bumper', 'buzz', 'cabin', 'cable', 'calf', 'california', 'camera',
       'cameras', 'candle', 'canidae', 'cannoli', 'cap', 'caption', 'car',
       'caravan'

In [7]:
# --- Task E: Topic Modeling (LDA) on Image Labels and Export Topic Words ---

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation
import pandas as pd
import numpy as np

# 1️⃣ Vectorize the image labels
vectorizer_lda = CountVectorizer(stop_words='english')
X_lda = vectorizer_lda.fit_transform(df_labeled['labels'].fillna(""))

# 2️⃣ Run LDA (you can adjust number of topics)
lda = LatentDirichletAllocation(n_components=5, random_state=42)
topic_matrix = lda.fit_transform(X_lda)

# 3️⃣ Add topic weights to the dataframe
topic_cols = [f"Topic_{i+1}" for i in range(lda.n_components)]
df_topics = pd.concat([df_labeled.reset_index(drop=True),
                       pd.DataFrame(topic_matrix, columns=topic_cols)], axis=1)

# 4️⃣ Sort by amount raised and split into top/bottom quartiles
df_sorted = df_topics.sort_values('amount_raised', ascending=False)
q1 = df_sorted['amount_raised'].quantile(0.25)
q3 = df_sorted['amount_raised'].quantile(0.75)
top_quartile = df_sorted[df_sorted['amount_raised'] >= q3]
bottom_quartile = df_sorted[df_sorted['amount_raised'] <= q1]

# 5️⃣ Compute average topic weights for each quartile
top_avg = top_quartile[topic_cols].mean()
bottom_avg = bottom_quartile[topic_cols].mean()

# 6️⃣ Create comparison table
topic_comparison = pd.DataFrame({
    'High $ (Top 25%)': top_avg,
    'Low $ (Bottom 25%)': bottom_avg
})
topic_comparison['Difference'] = topic_comparison['High $ (Top 25%)'] - topic_comparison['Low $ (Bottom 25%)']

print("\n📊 Average Topic Weights by Quartile:")
print(topic_comparison.round(3))

# 7️⃣ Extract top keywords for each topic and save to CSV
words = np.array(vectorizer_lda.get_feature_names_out())
topic_words = []

for i, topic in enumerate(lda.components_):
    top_word_list = [words[j] for j in topic.argsort()[-15:]]  # Top 15 words per topic
    topic_words.append({
        "Topic": f"Topic_{i+1}",
        "Top_Words": ", ".join(top_word_list)
    })
    print(f"\n🔹 Topic {i+1} top words:")
    print(top_word_list)

# Convert to DataFrame and save
df_topic_words = pd.DataFrame(topic_words)
df_topic_words.to_csv("lda_topic_words.csv", index=False)

print("\n✅ Topic words saved to 'lda_topic_words.csv'")



📊 Average Topic Weights by Quartile:
         High $ (Top 25%)  Low $ (Bottom 25%)  Difference
Topic_1             0.156               0.196      -0.039
Topic_2             0.262               0.217       0.045
Topic_3             0.300               0.213       0.087
Topic_4             0.152               0.195      -0.043
Topic_5             0.130               0.179      -0.049

🔹 Topic 1 top words:
['chair', 'handwriting', 'interior', 'facial', 'hairstyle', 'retail', 'text', 'finger', 'design', 'shelf', 'customer', 'furniture', 'restaurant', 'beauty', 'food']

🔹 Topic 2 top words:
['flooring', 'label', 'clip', 'pink', 'sign', 'red', 'wood', 'art', 'symbol', 'advertising', 'font', 'logo', 'graphic', 'graphics', 'design']

🔹 Topic 3 top words:
['forehead', 'sleeve', 'lips', 'shirt', 'vision', 'chin', 'glasses', 'eyewear', 'care', 'hair', 'eyebrow', 'expression', 'facial', 'happiness', 'smile']

🔹 Topic 4 top words:
['fictional', 'transport', 'mode', 'exterior', 'dog', 'arts', 'anim

In [8]:
df_topic_words

Topic                                          Top_Words
0  Topic_1  chair, handwriting, interior, facial, hairstyl...
1  Topic_2  flooring, label, clip, pink, sign, red, wood, ...
2  Topic_3  forehead, sleeve, lips, shirt, vision, chin, g...
3  Topic_4  fictional, transport, mode, exterior, dog, art...
4  Topic_5  lighting, parking, sunglasses, device, tail, b...

In [10]:
for i, topic in enumerate(lda.components_):
    top_word_list = [words[j] for j in topic.argsort()]  # Top 15 words per topic
    topic_words.append({
        "Topic": f"Topic_{i+1}",
        "Top_Words": ", ".join(top_word_list)
    })
    print(f"\n🔹 Topic {i+1} top words:")
    print(top_word_list)

# Convert to DataFrame and save
df_topic_words1 = pd.DataFrame(topic_words)


🔹 Topic 1 top words:
['electrical', 'cable', 'power', 'electricity', 'overhead', 'camera', 'cameras', 'optics', 'aviation', 'fixed', 'aerospace', 'brake', 'tail', 'digital', 'reflex', 'manufacturer', 'military', 'travel', 'air', 'engineering', 'operator', 'film', 'land', 'trucker', 'cricket', 'motor', 'mirror', 'driving', 'steering', 'rear', 'view', 'pizza', 'italian', 'cuisine', 'sun', 'leaf', 'secondary', 'institution', 'baseball', 'van', 'bumper', 'establishment', 'vegetable', 'rapper', 'window', 'town', 'facade', 'neighbourhood', 'area', 'urban', 'greenhouse', 'agriculture', 'plantation', 'woolen', 'wool', 'knit', 'branch', 'spring', 'grass', 'shoot', 'cosmetics', 'shadow', 'wig', 'earring', 'jewelry', 'barware', 'mode', 'transport', 'varnish', 'exterior', 'uniform', 'graffiti', 'tradesman', 'tool', 'soil', 'tire', 'chicken', 'comb', 'fowl', 'phasianidae', 'bird', 'goggles', 'chef', 'dishware', 'science', 'education', 'drink', 'stemware', 'alcoholic', 'spokesperson', 'government',

In [13]:
df_topic_words1.iloc[0]

Topic                                                  Topic_1
Top_Words    chair, handwriting, interior, facial, hairstyl...
Name: 0, dtype: object

In [14]:
topic_word_lists = []
for i, topic in enumerate(lda.components_):
    sorted_indices = topic.argsort()[::-1]          # sort words by probability (high → low)
    topic_words = [words[j] for j in sorted_indices] # all words in order
    topic_word_lists.append(topic_words)

# --- Make them into columns of equal length (pad shorter lists with blanks) ---
max_len = max(len(t) for t in topic_word_lists)
topic_columns = {}

for i, word_list in enumerate(topic_word_lists):
    padded_list = word_list + [""] * (max_len - len(word_list))
    topic_columns[f"Topic_{i+1}"] = padded_list

# --- Create DataFrame ---
df_topic_columns = pd.DataFrame(topic_columns)




In [15]:
df_topic_columns

Topic_1      Topic_2      Topic_3      Topic_4     Topic_5
0           food       design        smile      vehicle  automotive
1         beauty     graphics    happiness         blue       light
2     restaurant      graphic       facial   automotive        food
3      furniture         logo   expression         tire       night
4       customer         font      eyebrow        motor         cap
..           ...          ...          ...          ...         ...
578     overhead     overhead        cable        power        view
579  electricity        cable        power        cable        land
580        power        power  electricity     overhead       super
581        cable  electricity     overhead  electricity        bowl
582   electrical   electrical   electrical   electrical    american

[583 rows x 5 columns]

In [16]:
df_topic_columns.to_csv("lda_topic_words1.csv", index=False)

In [17]:
# --- Task E: Topic Modeling on Image Labels and Quartile Analysis ---

import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation

# 1️⃣ Prepare the text column
df_labeled['labels'] = df_labeled['labels'].fillna("")

# 2️⃣ Bag of Words for image labels
vectorizer = CountVectorizer(stop_words='english')
X = vectorizer.fit_transform(df_labeled['labels'])

# 3️⃣ Fit LDA model  (start with 5 topics – adjust if needed)
lda = LatentDirichletAllocation(n_components=5, random_state=42)
topic_matrix = lda.fit_transform(X)

# 4️⃣ Add topic weights to the dataframe
topic_cols = [f"Topic_{i+1}" for i in range(lda.n_components)]
df_topics = pd.concat([df_labeled.reset_index(drop=True),
                       pd.DataFrame(topic_matrix, columns=topic_cols)], axis=1)

# 5️⃣ Sort by likes (amount_raised or likes column)
df_sorted = df_topics.sort_values('amount_raised', ascending=False)

# 6️⃣ Compute quartiles
q1 = df_sorted['amount_raised'].quantile(0.25)
q3 = df_sorted['amount_raised'].quantile(0.75)

top_quartile = df_sorted[df_sorted['amount_raised'] >= q3]
bottom_quartile = df_sorted[df_sorted['amount_raised'] <= q1]

# 7️⃣ Average topic weights for each quartile
high_avg = top_quartile[topic_cols].mean()
low_avg = bottom_quartile[topic_cols].mean()

# 8️⃣ Combine results in one table
topic_comparison = pd.DataFrame({
    'High (Top 25%)': high_avg,
    'Low (Bottom 25%)': low_avg
})
topic_comparison['Difference'] = topic_comparison['High (Top 25%)'] - topic_comparison['Low (Bottom 25%)']

print("\n📊 Average Topic Weights by Quartile:")
print(topic_comparison.round(3))

# 9️⃣ Save both the full topic-weight matrix and summary table
df_topics.to_csv("TaskE_word_topic_matrix.csv", index=False)
topic_comparison.to_csv("TaskE_topic_quartile_summary.csv")

print("\n✅ Saved:")
print("• TaskE_word_topic_matrix.csv  → topic weights per image")
print("• TaskE_topic_quartile_summary.csv  → high vs low quartile comparison")



📊 Average Topic Weights by Quartile:
         High (Top 25%)  Low (Bottom 25%)  Difference
Topic_1           0.156             0.196      -0.039
Topic_2           0.262             0.217       0.045
Topic_3           0.300             0.213       0.087
Topic_4           0.152             0.195      -0.043
Topic_5           0.130             0.179      -0.049

✅ Saved:
• TaskE_word_topic_matrix.csv  → topic weights per image
• TaskE_topic_quartile_summary.csv  → high vs low quartile comparison


In [21]:
# --- Word-Topic Probability Matrix (for manual topic naming) ---

import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation

# 1️⃣ Prepare text data
df_labeled['labels'] = df_labeled['labels'].fillna("")

# 2️⃣ Bag of Words
vectorizer = CountVectorizer(stop_words='english')
X = vectorizer.fit_transform(df_labeled['labels'])
words = np.array(vectorizer.get_feature_names_out())

# 3️⃣ Fit LDA (4 topics in your case)
lda = LatentDirichletAllocation(n_components=5, random_state=42)
lda.fit(X)

# 4️⃣ Get the topic-word distribution matrix
# Shape = (n_topics, n_words)
topic_word_matrix = lda.components_

# 5️⃣ Normalize so that probabilities sum to 1 per topic
topic_word_probs = topic_word_matrix / topic_word_matrix.sum(axis=1)[:, np.newaxis]

# 6️⃣ Transpose → each row = word, each column = topic probability
word_topic_matrix = pd.DataFrame(topic_word_probs.T, columns=[f"Topic_{i+1}" for i in range(lda.n_components)])
word_topic_matrix.insert(0, "word", words)

# 7️⃣ Save to CSV
word_topic_matrix.to_csv("TaskE_word_topic_matrix2.csv", index=False)
print("✅ Saved as 'TaskE_word_topic_matrix1.csv'")
print("Columns: word, Topic_1, Topic_2, Topic_3, Topic_4")


✅ Saved as 'TaskE_word_topic_matrix1.csv'
Columns: word, Topic_1, Topic_2, Topic_3, Topic_4


In [22]:
# --- Task F: Generate amount_raised + topic weights table ---

import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation

# 1️⃣ Prepare your dataset
df_labeled['labels'] = df_labeled['labels'].fillna("")

# 2️⃣ Create Bag-of-Words matrix
vectorizer = CountVectorizer(stop_words='english')
X = vectorizer.fit_transform(df_labeled['labels'])

# 3️⃣ Fit LDA with 4 topics
lda = LatentDirichletAllocation(n_components=4, random_state=42)
topic_matrix = lda.fit_transform(X)   # each row = topic distribution for one fundraiser

# 4️⃣ Combine with amount_raised
topic_cols = [f"Topic{i+1}" for i in range(lda.n_components)]
df_topics = pd.concat([
    df_labeled[['amount_raised']].reset_index(drop=True),
    pd.DataFrame(topic_matrix, columns=topic_cols)
], axis=1)

# 5️⃣ Normalize topic weights to sum to 1 per row
df_topics[topic_cols] = df_topics[topic_cols].div(df_topics[topic_cols].sum(axis=1), axis=0)

# 6️⃣ Save and display
df_topics.to_csv("TaskF_amount_topic_weights.csv", index=False)

print("✅ Saved file: TaskF_amount_topic_weights.csv")
print(df_topics.head(10))   # show first 10 rows


✅ Saved file: TaskF_amount_topic_weights.csv
   amount_raised    Topic1    Topic2    Topic3    Topic4
0          18538  0.042223  0.041687  0.874380  0.041710
1          13541  0.035821  0.892740  0.035718  0.035722
2          35865  0.036672  0.036163  0.035732  0.891433
3          13939  0.035837  0.469087  0.035735  0.459342
4           5363  0.250000  0.250000  0.250000  0.250000
5           6214  0.036007  0.291177  0.319151  0.353665
6           8653  0.022732  0.930582  0.022729  0.023957
7           4988  0.812375  0.062604  0.062507  0.062514
8           4175  0.777231  0.032972  0.158519  0.031277
9           9443  0.904030  0.033425  0.031265  0.031281


In [5]:
# Topic Modeling and Quartile Comparison

import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation

# Prepare text data
df_labeled['labels'] = df_labeled['labels'].fillna("")

# Bag of Words
vectorizer = CountVectorizer(stop_words='english')
X = vectorizer.fit_transform(df_labeled['labels'])
words = np.array(vectorizer.get_feature_names_out())

# Fit LDA (4 topics)
lda = LatentDirichletAllocation(n_components=4, random_state=42)
topic_matrix = lda.fit_transform(X)   # document-topic distribution

# Get the topic-word distribution matrix
topic_word_matrix = lda.components_

# Normalize so that probabilities sum to 1 per topic
topic_word_probs = topic_word_matrix / topic_word_matrix.sum(axis=1)[:, np.newaxis]

# Transpose → each row = word, each column = topic probability
word_topic_matrix = pd.DataFrame(topic_word_probs.T,
                                 columns=[f"Topic_{i+1}" for i in range(lda.n_components)])
word_topic_matrix.insert(0, "word", words)

# Save word-topic matrix
word_topic_matrix.to_csv("word_topic_matrix_final.csv", index=False)

# Quartile comparison on amount raised / likes

# Combine topic weights with amount_raised
topic_cols = [f"Topic_{i+1}" for i in range(lda.n_components)]
df_topics = pd.concat([
    df_labeled[['amount_raised']].reset_index(drop=True),
    pd.DataFrame(topic_matrix, columns=topic_cols)
], axis=1)

# Normalize topic weights to sum to 1 per campaign
df_topics[topic_cols] = df_topics[topic_cols].div(df_topics[topic_cols].sum(axis=1), axis=0)

# Sort by actual amount raised
df_sorted = df_topics.sort_values('amount_raised', ascending=False)

# Define quartiles
q1 = df_sorted['amount_raised'].quantile(0.25)
q3 = df_sorted['amount_raised'].quantile(0.75)

# Split into top and bottom quartiles
top_quartile = df_sorted[df_sorted['amount_raised'] >= q3]
bottom_quartile = df_sorted[df_sorted['amount_raised'] <= q1]

# Calculate average topic weights for both quartiles
top_avg = top_quartile[topic_cols].mean()
bottom_avg = bottom_quartile[topic_cols].mean()

# Create comparison table
topic_comparison = pd.DataFrame({
    'High $ (Top 25%)': top_avg,
    'Low $ (Bottom 25%)': bottom_avg
})
topic_comparison['Difference'] = topic_comparison['High $ (Top 25%)'] - topic_comparison['Low $ (Bottom 25%)']
topic_comparison = topic_comparison.sort_values('Difference', ascending=False)

# Display
print("\n Average Topic Weights by Quartile:")
print(topic_comparison.round(3))


 Average Topic Weights by Quartile:
         High $ (Top 25%)  Low $ (Bottom 25%)  Difference
Topic_3             0.332               0.265       0.067
Topic_2             0.315               0.272       0.043
Topic_1             0.189               0.231      -0.042
Topic_4             0.164               0.232      -0.068
